Use SQLite3 DB to store the tasks

julmon opened the pull request.

julmon committed:
Do not apply_config() in bootstrap()

In BaseApplication class, we do not want anymore to load plugins
automatically when bootstrap() method is called because we'd need
to insert some work after the configuration is loaded and before
the plugins are loaded. This means apply_config() will need to be
executed explicilty by the app :

myapp.bootstrap()
...
myapp.apply_config()

julmon committed:
Use SQLite3 DB to store the tasks

This commit replaces the in-memory task list management by SQLite3
table.

dlax commented on tasklist/sqlite3_engine.py:
```diff
@@ -0,0 +1,314 @@
+                c.execute(
+                    dedent("""
+                        INSERT INTO tasks VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
+                    """),
```
The string would fit in a single line without triple quotes and dedent(), I think.

dlax commented on tasklist/sqlite3_engine.py:
```diff
@@ -0,0 +1,314 @@
+                    (
+                        task.id,
+                        task.worker_name,
+                        datetime_to_epoch(task.start_datetime),
+                        datetime_to_epoch(task.stop_datetime),
```
`datetime_to_epoch()` would return `0` if the argument if falsy, what's the meaning of storing 0 in the database (instead of `NULL`)?

julmon replied:
https://github.com/julmon/temboard-toolkit/blob/sqlite3/tasklist/sqlite3_engine.py#L264
https://github.com/julmon/temboard-toolkit/blob/sqlite3/tasklist/sqlite3_engine.py#L309

We need them to be not null for some queries.

julmon replied:
https://github.com/julmon/temboard-toolkit/blob/sqlite3/tasklist/sqlite3_engine.py#L264
https://github.com/julmon/temboard-toolkit/blob/sqlite3/tasklist/sqlite3_engine.py#L309

We need them to be not null for some queries.

dlax replied:
Ok, makes sense; thanks for the explanation.

dlax commented on tasklist/sqlite3_engine.py:
```diff
@@ -0,0 +1,314 @@
+                        datetime_to_epoch(task.start_datetime),
+                        datetime_to_epoch(task.stop_datetime),
+                        task.status,
+                        str(task.output) if task.output else '',
```
similarly, what's the meaning of storing an empty string instead of `NULL` here?

(Those two comments also apply to other methods below.)

dlax replied:
That one could be `NULL` (in contrast with date columns), isn't it?

dlax commented on tasklist/sqlite3_engine.py:
```diff
@@ -0,0 +1,314 @@
+                except ValueError as e:
+                    logger.exception(str(e))
+                    logger.debug(r)
+                    raise Exception("Could not decode task options.")
```
It'd be clearer if we tried to `json.loads(r[6])` independently of other operations in the `try` block (like `Task` instantiation). This way, we'd be certain to only catch `ValueError` raised by `json.loads()`.

dlax commented on tasklist/sqlite3_engine.py:
```diff
@@ -0,0 +1,314 @@
+                    dedent("""
+                        DELETE FROM tasks
+                        WHERE id = ?
+                    """),
```
Would also fit on a single line without triple quotes and dedent().

dlax commented on tasklist/sqlite3_engine.py:
```diff
@@ -0,0 +1,314 @@
+                            redo_interval=r[7],
+                            expire=r[8]
+                        )
+                    except ValueError as e:
+                        logger.exception(str(e))
+                        logger.error("Could not convert row to Task.")
+                        logger.debug(r)
```
Same here: it seems that `ValueError` should only concern `json.loads()`.

dlax commented on tests/test_sqlite3_engine.py:
```diff
@@ -0,0 +1,556 @@
+    engine = sqlite3_engine.TaskListSQLite3Engine(":memory:")
+    with pytest.raises(Exception) as e:
+        engine.insert(Task(id='cccc'))
+    assert "Could not insert task." in str(e.value)
```
I don't understand this part of the test, can you add a comment above? (Same for other test cases below.)

julmon replied:
Only for the sake of coverage, gonna remove them.

dlax commented on tasklist/sqlite3_engine.py:
```diff
@@ -0,0 +1,316 @@
+                )
+        except sqlite3.Error as e:
+            logger.exception(str(e))
+            raise Exception("Could not purge task list.")
```
I don't get the point of always catching `sqlite3.Error` and then raising a (less specific) `Exception`. Especially since the `Exception` is never catched in application code (only checked for in tests). On the other hand, logging the exception looks useful to me.

Also, since this involves a lot of repetition, it might be nice to factor this out into a context manager like:
```python
@contextmanager
def logged_cnx(conn):
    with conn:
       c = conn.cursor()
       try:
          yield c
       except sqlite3.Error as e:
          logger.exception(str(e))
          <re-raise or raise a specific exception type>
```

julmon replied:
Actually, I think we should be agnostic of storage engine in the upper layer: I don't want to deal with `sqlite3` errors in `taskmanager.py` because we could add support of another storage engine in the future, like Postgres. About less specific `Exception` raised, it could be replaced with a new `StorageEngineError` exception, what do you think ?

dlax replied:
I think that it does not matter much as long as the exception is not catched downstream. If it were to be catched, then I also agree that a specific and backend-agnostic exception type would be better.

julmon replied:
3c6dec1168179a1bff33d1381a380d906d2a4d2b should addresses it I hope.

julmon committed:
Vacuum DB once an hour

julmon committed:
Move schedule() and engine.vacuum() to serve1()

Because run() is deprecated and should be used only when testing the
framework.

julmon committed:
Set DB synchronous behavior to NORMAL

https://sqlite.org/pragma.html#pragma_synchronous

bersace commented on app.py:
```diff
@@ -97,7 +97,6 @@ def bootstrap(self, args, environ):
         config.add_specs(self.core_specs())
         config.add_specs(self.specs)
         config.load(**self.config_sources)
-        self.apply_config()
```
Pourquoi ?

bersace replied:
Faut mettre à jour le commentaire Stage 3.

bersace commented on tests/conftest.py:
```diff
@@ -0,0 +1,12 @@
+import pytest
+import os
+
+DBNAME = "/tmp/test.db"
```
Pourquoi ne pas utiliser systématiquement `:memory:` ?

dlax reviewed (commented):
This looks solid to me and well tested. Some questions and suggestions inline.

julmon committed:
Improvements

- Unit tests done with in-memory DB usage.
- Add StorageEngineError exception to handle backend errors.
- Minor improvements after code review.
